htdocs/libraries/icms/messaging/MentionService.php

```diff
@@ -0,0 +1,225 @@
+        $pattern = '/href\s*=\s*"[^"]*?uid=(\d+)/i';
+        preg_match_all($pattern, $content, $m);
+        if (!empty($m[1])) {
+            foreach ($m[1] as $uid) {
+                $uid = (int) $uid;
+                if ($uid > 0) {
+                    $uids[$uid] = true;
+                }
+            }
+        }
+        return array_map('intval', array_keys($uids));
+    }
+
+    /**
+     * Map usernames to user IDs for active users only.
+     * Returns list of ints.
+     */
+    private static function mapUsernamesToIds(array $usernames): array
+    {
+        if (empty($usernames)) {
+            return [];
+        }
+
+        // Build placeholders
+        $placeholders = [];
+        $params = [];
+        foreach ($usernames as $i => $uname) {
+            $key = ":u{$i}";
+            $placeholders[] = $key;
+            $params[$key] = $uname;
+        }
+
+        try {
+            /** @var PDO $pdo */
+            $pdo = \icms::$db;
+            $sql = 'SELECT uid FROM ' . str_replace('`','', \icms::$xoopsDB->prefix('users')) .
+                ' WHERE level > 0 AND LOWER(uname) IN (' . implode(',', $placeholders) . ')';
+            $stmt = $pdo->prepare($sql);
+            foreach ($params as $k => $v) {
+                $stmt->bindValue($k, strtolower($v), PDO::PARAM_STR);
+            }
+            $stmt->execute();
+            $uids = [];
+            while ($row = $stmt->fetch(PDO::FETCH_ASSOC)) {
+                $uids[] = (int) $row['uid'];
+            }
+            return $uids;
+        } catch (Throwable $e) {
+            return [];
+        }
+    }
+
+    /**
+     * Ensure target user IDs are active.
+     */
+    private static function filterActiveUsers(array $userIds): array
+    {
+        if (empty($userIds)) {
+            return [];
+        }
+        $placeholders = [];
+        $params = [];
+        foreach ($userIds as $i => $uid) {
+            $key = ":id{$i}";
+            $placeholders[] = $key;
+            $params[$key] = (int) $uid;
+        }
+        try {
+            /** @var PDO $pdo */
+            $pdo = \icms::$db;
+            $sql = 'SELECT uid FROM ' . str_replace('`','', \icms::$xoopsDB->prefix('users')) .
+                ' WHERE level > 0 AND uid IN (' . implode(',', $placeholders) . ')';
+            $stmt = $pdo->prepare($sql);
+            foreach ($params as $k => $v) {
+                $stmt->bindValue($k, $v, PDO::PARAM_INT);
+            }
+            $stmt->execute();
+            $uids = [];
+            while ($row = $stmt->fetch(PDO::FETCH_ASSOC)) {
+                $uids[] = (int) $row['uid'];
+            }
+            return $uids;
+        } catch (Throwable $e) {
+            return [];
+        }
+    }
+
+    private static function buildBody(string $subject, string $url, string $senderName, string $context): string
+    {
+        $site = (string) ($GLOBALS['icmsConfig']['sitename'] ?? 'ImpressCMS');
```

Comment:
I would prefer more using `icms::$config` instead of $GLOBALS, because they make code much harder testable